### PR TITLE
[RLlib] Adding Epistemic Neural Network Example

### PR DESCRIPTION
**Pull request by tlaurie99**
## Why are these changes needed?

This PR provides examples on how to provide:

#### Learner
- How to build a custom learner with a custom loss function.
- Accessing custom information in `batch`.
- Compute losses with actor, critic, and epinet networks using TD loss.

#### RL module
- Builds a custom RL module with custom `_forward` methods.
- Specify the `_forward_train` dictionary output to include needed keys and custom keys.
- A custom `build_network_layers` method demonstrating how to build layers modularly.
- Adding components to `batch`.

#### Loss Function

- Computes a custom loss function use base PPO's surrogate loss calculation with the critic loss including part of the epistemic neural network.

## Related issue number
https://discuss.ray.io/t/contributing-utilities-to-rllib/21320

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/learners/classes/epinet_rlm.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        enn_output = learnable + prior
+        return {
+            CRITIC_OUTPUT_BASE: base_output,
+            CRITIC_OUTPUT_ENN: enn_output,
+        }
+
+    @override(RLModule)
+    def _forward_train(self, batch: Dict[str, Any]) -> Dict[str, Any]:
+        obs = batch[Columns.OBS]
+        if Columns.NEXT_OBS not in batch:
```
Comment (tlaurie99) [inline rllib/examples/learners/classes/epinet_rlm.py:244]: @simonsays1980 I am still have issues with the `new_obs` being in the `batch`. I've tried using a different pipeline in the config and after about three hours of looking I think it is for the dummy batch initialization. After the batch is initialized, I see that `new_obs` is always in the `batch`, but at the end of an iteration, or when initializing, it is not.

File: rllib/examples/learners/epinet.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        .resources(num_gpus=1)
+        .env_runners(preprocessor_pref="rllib")
+        .reporting(keep_per_episode_custom_metrics=True)
+    )
+
+algo = base_config.build()
+total_timesteps = 0
+default_timesteps = 200_000
+default_reward = 250
+for iteration in range(1000):
```
Comment (tlaurie99) [inline rllib/examples/learners/epinet.py:182]: @simonsays1980 I also had to revert to using this old way of training as the `run_rllib_example_script_experiment` was not using my config or my rl module.
